src/backend/distributed/transaction/transaction_management.c

```diff
@@ -75,11 +148,25 @@ CoordinatedTransactionCallback(XactEvent event, void *arg)
 			Assert(!subXactAbortAttempted);
 			CurrentCoordinatedTransactionState = COORD_TRANS_NONE;
 			XactModificationLevel = XACT_MODIFICATION_NONE;
+			dlist_init(&InProgressTransactions);
+			CurrentTransactionUse2PC = false;
 		}
 		break;
 
 		case XACT_EVENT_ABORT:
 		{
+			/*
+			 * FIXME: Add warning for the COORD_TRANS_COMMITTED case. That
+			 * can be reached if this backend fails after the
+			 * XACT_EVENT_PRE_COMMIT state.
+			 */
+
+			/* handles both already prepared and open transactions */
+			if (CurrentCoordinatedTransactionState > COORD_TRANS_IDLE)
+			{
+				CoordinatedRemoteTransactionsAbort();
+			}
+
 			/* close connections etc. */
 			if (CurrentCoordinatedTransactionState != COORD_TRANS_NONE)
```

Comment:
I'd rather make sure that we reset per-transaction state in any case.  Not that important though, if you feel strongly.